vimiv/imutils/exif.py

```diff
@@ -11,138 +11,267 @@
+                                    exif[keyname] = (keyname, val)
+                                elif keytype in (
+                                    piexif.TYPES.Ascii,
+                                    piexif.TYPES.Undefined,
+                                ):  # byte encoded
+                                    exif[keyname] = (keyname, val.decode())
+                                elif keytype in (
+                                    piexif.TYPES.Rational,
+                                    piexif.TYPES.SRational,
+                                ):  # (int, int) <=> numerator, denominator
+                                    exif[keyname] = (keyname, f"{val[0]}/{val[1]}")
+
+                    except (piexif.InvalidImageDataError, KeyError):
+                        return None
+
+                    return exif
+
+                def copy_exif(self, dest: str, reset_orientation: bool = True) -> None:
+                    """Copy exif information from current image to dest.
+
+                    Args:
+                        dest: Path to write the exif information to.
+                        reset_orientation: If true, reset the exif orientation tag to
+                            normal.
+                    """
+
+                    try:
+                        if reset_orientation:
+                            with contextlib.suppress(KeyError):
+                                self._metadata["0th"][
+                                    piexif.ImageIFD.Orientation
+                                ] = ExifOrientation.Normal
+                        exif_bytes = piexif.dump(self._metadata)
+                        piexif.insert(exif_bytes, dest)
+                        _logger.debug("Successfully wrote exif data for '%s'", dest)
+                    except piexif.InvalidImageDataError:  # File is not a jpg
+                        _logger.debug(
+                            "File format for '%s' does not support exif", dest
+                        )
+                    except ValueError:
+                        _logger.debug("No exif data in '%s'", dest)
+
+                def exif_date_time(self) -> str:
+                    """Exif creation date and time of filename."""
+
+                    with contextlib.suppress(
+                        piexif.InvalidImageDataError, FileNotFoundError, KeyError
+                    ):
+                        return self._metadata["0th"][piexif.ImageIFD.DateTime].decode()
+                    return ""
+
+            return ExifHandlerPiexif
+
+        if return_value is not None:
             return return_value
 
-        if piexif is None:
-            return stub
-        return func
+        class NoExifHandler:
+            """Handler for exif access without hander library."""
 
-    return decorator
+            def __init__(self, *args):
+                _logger.debug("No exif support.")
 
+            def __getattribute__(self, name):
```

Comment:
As much as I like avoiding code duplication, currently this does not work properly as the `lambda` always returns `None`. This crashes on `exif_date_time` with a `TypeError` as we expect a string, here the empty string, as return value.

I guess `get_formatted_exif` would die in a similar fashion as we always expect a `dict`, but this is currently not called anywhere without exif support.

I see two options here:
- Implement the functions explicitly and return the proper type after logging / messaging.
- Make the lambda raise an exception, ideally a new custom exception like `NoExifSupport`, and enforce the caller to deal with it explicitly instead of checking for `... is None ...`. Like this the caller can think about messaging or the return type. The code in `filelist.exif_date_time` would then read:
```
@api.status.module("{exif-date-time}")
def exif_date_time() -> str:
    """Exif creation date and time of the current image.

    This is meant as an example api.status.module to show how to display exif
    data in the statusbar. If there are any requests/ideas for more, this can
    be used as basis to work with.
    """
    try:
        return imutils.exif.ExifHandler(current()).exif_date_time()
    except imutils.exif.NoExifSupport:
        return ""
```
and this `__getattribute__` could be
```
            def __getattribute__(self, name):
                if name in ["copy_exif", "exif_date_time", "get_formatted_exif"]:
                    raise NoExifSupport()
                return super().__getattribute__(name)
```

I somehow feel like the exception version could be the nicer, and more extendible, version. What do you think?

Reply:
I see, my idea was probably not yet well-conceived :D I think both of your ideas are reasonable. If we implement the functions explicitly, we could use an abstract class the to make it more obvious that all three classes have the same functionality, as mentioned above. I also quite like the idea of of throwing a custom exception. Probably, this would be the more pyhtonic solutions.

Reply:
I have decided to use a abstract class `_ExifHandler` and define three implementations of this class. This leas to more code compared to the  `__getattribute__` based implementations but it seems to be cleaner and less *hacked*. But to get the best of both sides, I have also added the `NoExifSupport` custom exception.

Implemented in 4fd97f9f4bdb73e583092ab49486144d4a08fb36 and 4fd97f9f4bdb73e583092ab49486144d4a08fb36.

Reply:
Although I have no issue with a funky `__getattribute__` implementation this is certainly cleaner, nice!